hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapContainer.java

```diff
@@ -161,26 +175,26 @@ public boolean containsValue(boolean binary, Data value) {
     }
 
     public int size() {
-        int size = 0;
-        for (MultiMapValue multiMapValue : multiMapValues.values()) {
-            size += multiMapValue.getCollection(false).size();
-        }
-        return size;
+        return size.get();
     }
 
     public int clear() {
+        int numOfEntriesDeleted = 0;
+        int numOfValuesDeleted = 0;
         Collection<Data> locks = lockStore != null ? lockStore.getLockedKeys() : Collections.<Data>emptySet();
-        Map<Data, MultiMapValue> lockedKeys = createHashMap(locks.size());
-        for (Data key : locks) {
-            MultiMapValue multiMapValue = multiMapValues.get(key);
-            if (multiMapValue != null) {
-                lockedKeys.put(key, multiMapValue);
+
+        for (Data key : multiMapValues.keySet()) {
+            if (!locks.contains(key)) {
+                MultiMapValue value = multiMapValues.remove(key);
+                if (value != null) {
+                    numOfEntriesDeleted++;
+                    numOfValuesDeleted += value.size();
+                }
             }
         }
```

Comment:
Due to the threading model, no other operation on this MultiMapContainer can be executed. So you don't need to worry about e.g. concurrent adds in combination with clean.

I need to think about increased complexity for the clear.

Reply:
I just made a quick hack to demonstrate how it can be complexity can be reduced.

```
 public int clear() {
        Collection<Data> locks = lockStore != null ? lockStore.getLockedKeys() : Collections.<Data>emptySet();
        Map<Data, MultiMapValue> lockedKeys = createHashMap(locks.size());
        int lockedValueCount = 0;
        for (Data key : locks) {
            MultiMapValue multiMapValue = multiMapValues.get(key);
            if (multiMapValue != null) {
                lockedKeys.put(key, multiMapValue);
                lockedValueCount+=multiMapValue.getCollection(false).size();
            }
        }
        int numberOfAffectedEntries = multiMapValues.size() - lockedKeys.size();
        multiMapValues.clear();
        multiMapValues.putAll(lockedKeys);
        size=lockedValueCount;
        return numberOfAffectedEntries;
    }
```